src/collections/deque.rs

```diff
@@ -81,29 +94,182 @@ impl<T: State<S>, S: Write> Deque<T, S> {
         Ok(())
     }
 
-    pub fn pop_front(&mut self) -> Result<()> {
-        if self.len() == 0 {
-            return Ok(());
+    pub fn pop_front(&mut self) -> Result<Option<ReadOnly<T>>> {
+        if self.is_empty() {
+            return Ok(None);
         }
 
-        self.map.remove(self.meta.head);
         self.meta.head += 1;
-        
-        Ok(())
+        self.map.remove(self.meta.head - 1)
     }
 
-    pub fn pop_back(&mut self) -> Result<()> {
-        if self.len() == 0 {
-            return Ok(());
+    pub fn pop_back(&mut self) -> Result<Option<ReadOnly<T>>> {
+        if self.is_empty() {
+            return Ok(None);
         }
 
         self.meta.tail -= 1;
-        self.map.remove(self.meta.tail);
-        
-        Ok(())
+        self.map.remove(self.meta.tail)
     }
 
     pub fn get_mut(&mut self, index: u64) -> Result<Option<ChildMut<u64, T, S>>> {
         self.map.get_mut(index + self.meta.head)
     }
 }
+
+#[allow(unused_imports)]
+mod test {
+    use super::*;
+    use crate::store::MapStore;
+
+    #[test]
+    fn deque_u32_create() {
+        let store = Store::new(MapStore::new());
+        let _deque: Deque<u32> = Deque::create(store.clone(), Meta::default()).unwrap();
+    }
+
+    #[test]
+    fn deque_u32_push_front() {
+        let store = Store::new(MapStore::new());
+        let mut deque: Deque<u32> = Deque::create(store.clone(), Meta::default()).unwrap();
+
+        deque.push_front(42).unwrap();
+    }
+
+    #[test]
+    fn deque_u32_push_back() {
+        let store = Store::new(MapStore::new());
+        let mut deque: Deque<u32> = Deque::create(store.clone(), Meta::default()).unwrap();
+
+        deque.push_back(42).unwrap();
+    }
+
+    #[test]
+    fn deque_u32_pop_front_empty() {
+        let store = Store::new(MapStore::new());
+        let mut deque: Deque<u32> = Deque::create(store.clone(), Meta::default()).unwrap();
+
+        match deque.pop_front().unwrap() {
+            Some(_) => assert!(false),
+            None => (),
+        }
+    }
+
+    #[test]
+    fn deque_u32_pop_front() {
+        let store = Store::new(MapStore::new());
+        let mut deque: Deque<u32> = Deque::create(store.clone(), Meta::default()).unwrap();
+
+        deque.push_front(42).unwrap();
+
+        match deque.pop_front().unwrap() {
+            Some(v) => assert_eq!(*v, 42),
+            None => (),
```

Comment:
If `pop_front` had a bug where it mistakenly returned `None`, wouldn't this test still pass?

Reply:
(I think a similar issue applies to a few of the other test cases)

Reply:
Yes, good point. I had a better pattern in the pop_back tests, but forgot to adjust this one to match that. Also added is_empty assertions after the pop on both.